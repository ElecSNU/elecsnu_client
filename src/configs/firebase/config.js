import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/storage';

const firebaseConfig = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId:
        process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_APP_ID,
};

const fire_app = firebase.initializeApp(firebaseConfig);

export const fire_auth = fire_app.auth();
export const fire_store = fire_app.firestore();
export const fire_storage = fire_app.storage();
export default fire_app;
