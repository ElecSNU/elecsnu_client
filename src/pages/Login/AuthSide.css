#auth-side {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    overflow: hidden;
}

#with-pass,
#password-less {
    height: 100%;
    width: 50%;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    align-items: center;
}

#with-pass .login-heading {
    font-size: 2rem;
}

#with-pass #get-input {
    display: flex;
    flex-direction: column;
    height: 15%;
    width: 50%;
    justify-content: space-between;
}

#with-pass {
    transform: translate(0%, 0);
    opacity: 1;
    transition: all 0.8s ease-out;
    pointer-events: all;
}

#with-pass.show-password-less {
    transform: translate(-100%, 0);
    opacity: 0;
    pointer-events: none;
    /* background-color: var(--dark); */
}

.submit-container {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    position: relative;
}

.passwordless-option {
    color: var(--accent2);
    font-size: 1.5rem;
    cursor: pointer;
    margin-top: 1.5rem;
    transition: all 0.6s;
}

.passwordless-option:hover {
    color: var(--background);
}
