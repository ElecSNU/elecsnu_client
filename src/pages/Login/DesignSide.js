import React from 'react';

const DesignSide = () => {
    return (
        <section
            id='design-side'
            className='background-1'
        ></section>
    );
};

export default DesignSide;
