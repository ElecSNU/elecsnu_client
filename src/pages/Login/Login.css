#login {
    display: flex;
    height: 100%;
}

#design-side,
#auth-side {
    height: 100%;
    flex: 1;
}

.input-text {
    padding: 1rem 1.5rem;
    width: 100%;
    font-size: 1rem;
    border-radius: 5rem;
    border: transparent;
    transition: all 0.7s;
    background-color: rgba(0, 0, 0, 0.1);
}

.input-text:focus {
    border: 2px solid var(--background);
}
