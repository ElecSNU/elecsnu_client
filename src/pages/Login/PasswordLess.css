#password-less {
    height: 100%;
    width: 50%;
    background-color: rgba(255, 255, 255, 0.8);
    transform: translate(100%, 0);
    opacity: 0;
    pointer-events: none;
    /* transition: transform 0.7s ease-out; */
}

#password-less.show-password-less {
    transform: translate(0%, 0);
    opacity: 1;
    animation: show 0.8s ease-out;
    pointer-events: all;
}

@keyframes show {
    from {
        opacity: 0;
        transform: translate(100%, 0);
    }

    to {
        opacity: 1;
        transform: translate(0%, 0);
    }
}

#id-image {
    width: 30%;
    height: 45%;
    border: 2px solid var(--accent2);
}

#id-image img {
    width: 100%;
    height: 100%;
}
