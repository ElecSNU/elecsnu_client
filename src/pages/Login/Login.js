import React, { useEffect, useState } from 'react';
import { fire_auth } from '../../configs/firebase/config';
import { toast } from 'react-toastify';

import './Login.css';

const DesignSide = React.lazy(() => import('./DesignSide'));
const AuthSide = React.lazy(() => import('./AuthSide'));

const Login = () => {
    useEffect(() => {
        let urlParams = new URLSearchParams(
            window.location.search
        );

        let email = '';

        if (urlParams.get('apiKey')) {
            if (
                fire_auth.isSignInWithEmailLink(
                    window.location.href
                )
            ) {
                email = window.localStorage.getItem(
                    'emailForSignIn'
                );
                if (!email) {
                    email = window.prompt(
                        'Please provide your email for confirmation'
                    );
                }
                fire_auth
                    .signInWithEmailLink(
                        email,
                        window.location.href
                    )
                    .then((result) => {
                        console.log(result);
                        window.localStorage.removeItem(
                            'emailForSignIn'
                        );
                        toast.success(
                            'Logged in Successfully'
                        );
                    })
                    .catch((error) => {
                        console.error(error);
                        toast.error(error.message);
                    });
            }
        }
    }, []);

    const [
        showPasswordless,
        setShowPasswordless,
    ] = useState(false);

    return (
        <div
            id='login'
            className='border-radius-15 box-shadow'
        >
            <DesignSide
                showPasswordless={showPasswordless}
            />
            <AuthSide
                showPasswordless={showPasswordless}
                setShowPasswordless={setShowPasswordless}
            />
        </div>
    );
};

export default Login;
