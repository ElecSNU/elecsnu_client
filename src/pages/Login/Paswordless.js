import React, { useRef, useState } from 'react';
import Quagga from 'quagga';
import { toast } from 'react-toastify';
import { useStoreActions } from 'easy-peasy';

import useFirestore from '../../hooks/useFirstore';

import './PasswordLess.css';

const Paswordless = ({
    showPasswordless,
    setShowPasswordless,
}) => {
    const fileRef = useRef();

    const voters_data = useFirestore('voters');

    const [idCardImage, setIdCardImage] = useState('');
    const [detectedRoll, setDetectedRoll] = useState(null);

    const passwordless_login_action = useStoreActions(
        (actions) => actions.accountModel.passwordless_login
    );

    const scan = async (img) => {
        await Quagga.decodeSingle(
            {
                src: img,
                numOfWorkers: 0, // Needs to be 0 when used within node
                locate: true,
                decoder: {
                    readers: ['code_128_reader'], // List of active readers
                },
            },
            (result) => {
                if (result && result.codeResult) {
                    console.log(result);
                    setDetectedRoll(result.codeResult.code);
                } else {
                    console.log('not detected', result);
                }
            }
        );
    };

    const loginPasswordless = async () => {
        if (detectedRoll != null) {
            let voter_found = false;
            let voterEmail = '';
            voters_data.docs.forEach((doc) => {
                if (doc.id === detectedRoll) {
                    voter_found = true;
                    voterEmail = `${doc.net_id}@snu.edu.in`;
                }
            });

            if (!voter_found || voterEmail === '') {
                toast.error('User not Found!');
            } else {
                passwordless_login_action({
                    email: voterEmail,
                });
            }
        } else {
            toast.error(
                'Roll Number not detected from ID Card! Please select a valid image'
            );
        }
    };

    return (
        <div
            id='password-less'
            className={
                showPasswordless ? 'show-password-less' : ''
            }
        >
            <h1
                className='login-heading'
                style={{ fontSize: '2rem' }}
            >
                LOGIN
            </h1>

            <div className='submit-container'>
                <input
                    type='file'
                    accept='image/*'
                    style={{
                        display: 'none',
                        opacity: '0',
                        width: '100%',
                        height: '100%',
                        position: 'absolute',
                    }}
                    ref={fileRef}
                    onChange={async () => {
                        URL.revokeObjectURL(idCardImage);

                        if (fileRef.current.files[0]) {
                            console.log(
                                fileRef.current.files[0]
                            );
                            let imgURL = URL.createObjectURL(
                                fileRef.current.files[0]
                            );
                            setIdCardImage(imgURL);
                            await scan(imgURL);
                        }
                    }}
                />
                <button
                    className='submit-border dark'
                    onClick={() => {
                        fileRef.current.click();
                    }}
                >
                    Choose ID Card image
                </button>
            </div>
            <div
                id='id-image'
                style={{
                    height:
                        idCardImage === '' ? '0' : '45%',
                    border:
                        idCardImage === ''
                            ? 'none'
                            : '2px solid var(--accent2)',
                }}
            >
                <img
                    src={idCardImage}
                    alt=''
                    style={{
                        display:
                            idCardImage === ''
                                ? 'none'
                                : 'block',
                    }}
                />
            </div>
            <div className='submit-container'>
                <button
                    className='submit-button'
                    onClick={async () => {
                        loginPasswordless();
                    }}
                >
                    LET'S VOTE
                </button>
                <p
                    className='passwordless-option'
                    onClick={() => {
                        setIdCardImage('');
                        setDetectedRoll(null);
                        setShowPasswordless(false);
                    }}
                >
                    Sign in with Email and Password
                </p>
            </div>
        </div>
    );
};

export default Paswordless;
