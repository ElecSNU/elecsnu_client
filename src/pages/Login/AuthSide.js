import React, { useRef, useEffect, useState } from 'react';
import { useStoreActions, useStoreState } from 'easy-peasy';
import { toast } from 'react-toastify';
import { Redirect } from 'react-router-dom';
import useLoader from '../../hooks/useLoader';

import './AuthSide.css';

const PasswordLess = React.lazy(() =>
    import('./Paswordless')
);

const AuthSide = ({
    showPasswordless,
    setShowPasswordless,
}) => {
    const showLoader = useLoader();

    const loginStatus = useStoreState(
        (store) => store.accountModel.temp_status
    );

    const [loggedIn, setLoggedIn] = useState(false);

    const loginAction = useStoreActions(
        (actions) => actions.accountModel.login
    );
    const checkLoginAction = useStoreActions(
        (actions) => actions.accountModel.check_login
    );

    const email = useRef();
    const pass = useRef();

    const handleLogin = async () => {
        showLoader(true);
        loginAction({
            email: email.current.value,
            password: pass.current.value,
        });
    };

    useEffect(() => {
        const unsubscribe = checkLoginAction();
        showLoader(false);

        if (loginStatus != null) {
            console.log('hey');

            if (loginStatus[0]) {
                toast.success(loginStatus[1]);
                setLoggedIn(true);

                window.location.pathname = '/';
            } else {
                toast.error(loginStatus[1]);
                setLoggedIn(false);
            }
        }

        return unsubscribe;

        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [loginStatus]);

    // if (loggedIn) {
    //     return <Redirect to='/' />;
    // } else
    return (
        <section id='auth-side'>
            <PasswordLess
                showPasswordless={showPasswordless}
                setShowPasswordless={setShowPasswordless}
            />
            <div
                id='with-pass'
                className={
                    showPasswordless
                        ? 'show-password-less'
                        : ''
                }
            >
                <h1 className='login-heading'>LOGIN</h1>
                <div id='get-input'>
                    <input
                        className='input-text'
                        type='text'
                        ref={email}
                        placeholder='Enter SNU Net ID'
                    />
                    <input
                        className='input-text'
                        type='password'
                        ref={pass}
                        placeholder='Enter password'
                    />
                </div>
                <div className='submit-container'>
                    <button
                        className='submit-button'
                        onClick={handleLogin}
                    >
                        LET'S VOTE
                    </button>
                    <p
                        className='passwordless-option'
                        onClick={() =>
                            setShowPasswordless(true)
                        }
                    >
                        Sign in with ID Card
                    </p>
                </div>
            </div>
        </section>
    );
};

export default AuthSide;
