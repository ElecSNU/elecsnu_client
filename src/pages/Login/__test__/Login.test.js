import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import ReactDOM from 'react-dom'
// import React , { useState }from 'react'
import renderer from 'react-test-renderer'
import DesignSide from './../DesignSide'
import AuthSide from './../AuthSide'
import Login from './../Login'
import React, { useRef, useEffect, useState } from 'react';
import { useStoreActions, useStoreState } from 'easy-peasy';
import Hooks from './../../../hooks/useFirstore'
// import { useState, useEffect } from 'react';
import { fire_store } from './../../../configs/firebase/config';


afterEach(cleanup)
test('Tests the rendering of the Design side', () => {
    const div = document.createElement("div")
    const tree = renderer.create(<DesignSide></DesignSide>).toJSON();
    expect(tree).toMatchSnapshot();
   });
   
// test('Tests the rendering of the Authentication side', () => {
//     const div = document.createElement("div")
//     const tree = renderer.create(<AuthSide showPasswordless={false} setShowPasswordless={() => {}} ></AuthSide>).toJSON();
//     expect(tree).toMatchSnapshot();
//    });   

// test('Tests the rendering of whole login', () => {
//     const div = document.createElement("div")
//     const DesignSide = React.lazy(() => import('./../DesignSide'));
//     const AuthSide = React.lazy(() => import('./../AuthSide'));
//     const tree = renderer.create(<Login></Login>).toJSON();
//     expect(tree).toMatchSnapshot();
// });   