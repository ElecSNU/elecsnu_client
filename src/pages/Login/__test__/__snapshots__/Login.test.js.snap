// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests the rendering of the Authentication side 1`] = `
<section
  className="background-1 "
  id="design-side"
>
  <div
    className="text-1 heading-text foreground-light"
  >
    <div>
      OUR
    </div>
    <div>
      VOTE
    </div>
  </div>
  <div
    className="text-2 heading-text foreground-light"
  >
    <div>
      IS
    </div>
    <div>
      OUR
    </div>
    <div>
      VOICE
    </div>
  </div>
</section>
`;

exports[`Tests the rendering of the Design side 1`] = `
<section
  className="background-1 "
  id="design-side"
>
  <div
    className="text-1 heading-text foreground-light"
  >
    <div>
      OUR
    </div>
    <div>
      VOTE
    </div>
  </div>
  <div
    className="text-2 heading-text foreground-light"
  >
    <div>
      IS
    </div>
    <div>
      OUR
    </div>
    <div>
      VOICE
    </div>
  </div>
</section>
`;
